### ISSUE TITLE
EKS Cluster Launch Failure - Invalid Base 64 User Encoding of user data.

### ISSUE THREAD

**Issue by haydursek** [rancher/rancher#31612]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
Launch EKS cluster using RKE into AWS, using default option and removing all user data present in the UI. Interestingly I have now gotten this error when leaving in the default user data which is present in the UI and when removing it. 

**Result:**
Cluster fails to launch with the following message:
` InvalidUserData.Malformed: Invalid BASE64 encoding of user data.\n\tstatus code: 400, request id: 9079e182-f0a5-43b1-a5b2-56a56eed1f19, requeuing"`

**Other details that may be helpful:**
Below is the request sent from the Rancher UI:
```
{
   "actions":{
      "enableMonitoring":"…/v3/clusters/c-8bpkc?action=enableMonitoring",
      "generateKubeconfig":"…/v3/clusters/c-8bpkc?action=generateKubeconfig",
      "importYaml":"…/v3/clusters/c-8bpkc?action=importYaml"
   },
   "agentImage":"",
   "agentImageOverride":"",
   "allocatable":{
      "cpu":"0",
      "memory":"0",
      "pods":"0"
   },
   "annotations":{
      "authz.management.cattle.io/creator-role-bindings":"{\"created\":[\"cluster-owner\"],\"required\":[\"cluster-owner\"]}",
      "lifecycle.cattle.io/create.cluster-agent-controller-cleanup":"true",
      "lifecycle.cattle.io/create.cluster-provisioner-controller":"true",
      "lifecycle.cattle.io/create.cluster-scoped-gc":"true",
      "lifecycle.cattle.io/create.mgmt-cluster-rbac-remove":"true"
   },
   "answers":{
      "clusterId":null,
      "projectId":null,
      "type":"/v3/schemas/answer"
   },
   "appliedEnableNetworkPolicy":false,
   "appliedPodSecurityPolicyTemplateId":"",
   "appliedSpec":{
      "agentImageOverride":"",
      "answers":{
         "clusterId":null,
         "projectId":null,
         "type":"/v3/schemas/answer"
      },
      "clusterTemplateId":null,
      "clusterTemplateRevisionId":null,
      "defaultClusterRoleForProjectMembers":null,
      "defaultPodSecurityPolicyTemplateId":null,
      "description":"",
      "desiredAgentImage":"",
      "desiredAuthImage":"",
      "displayName":"",
      "dockerRootDir":"/var/lib/docker",
      "enableClusterAlerting":false,
      "enableClusterMonitoring":false,
      "enableNetworkPolicy":null,
      "internal":false,
      "localClusterAuthEndpoint":{
         "enabled":false,
         "type":"/v3/schemas/localClusterAuthEndpoint"
      },
      "type":"/v3/schemas/clusterSpec",
      "windowsPreferedCluster":false
   },
   "authImage":"",
   "baseType":"cluster",
   "capabilities":{
      "loadBalancerCapabilities":{
         "healthCheckSupported":false,
         "type":"/v3/schemas/loadBalancerCapabilities"
      },
      "nodePoolScalingSupported":false,
      "pspEnabled":false,
      "type":"/v3/schemas/capabilities"
   },
   "capacity":{
      "cpu":"0",
      "memory":"0",
      "pods":"0"
   },
   "clusterTemplateId":null,
   "clusterTemplateRevisionId":null,
   "conditions":[
      11 items{
         "status":"True",
         "type":"Pending"
      },
      {
         "lastUpdateTime":"2021-03-06T16:32:36Z",
         "status":"True",
         "type":"Provisioned"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:34Z",
         "message":"Waiting for API to be available",
         "status":"Unknown",
         "type":"Waiting"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:33Z",
         "status":"True",
         "type":"BackingNamespaceCreated"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:33Z",
         "status":"True",
         "type":"DefaultProjectCreated"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:33Z",
         "status":"True",
         "type":"SystemProjectCreated"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:33Z",
         "status":"True",
         "type":"InitialRolesPopulated"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:34Z",
         "status":"True",
         "type":"CreatorMadeOwner"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:34Z",
         "status":"True",
         "type":"NoDiskPressure"
      },
      {
         "lastUpdateTime":"2021-03-06T16:18:34Z",
         "status":"True",
         "type":"NoMemoryPressure"
      },
      {
         "lastUpdateTime":"2021-03-06T16:37:41Z",
         "message":"InvalidUserData.Malformed: Invalid BASE64 encoding of user data.\n\tstatus code: 400, request id: 7c8c9d45-cd42-4264-b8b6-acee9669557f",
         "status":"False",
         "type":"Updated"
      }
   ],
   "created":"2021-03-06T16:18:33Z",
   "createdTS":1615047513000,
   "creatorId":"user-7tb4l",
   "defaultClusterRoleForProjectMembers":null,
   "defaultPodSecurityPolicyTemplateId":null,
   "description":"",
   "desiredAgentImage":"",
   "desiredAuthImage":"",
   "dockerRootDir":"/var/lib/docker",
   "driver":"EKS",
   "eksConfig":{
      "amazonCredentialSecret":"cattle-global-data:cc-twprg",
      "displayName":"test-cluster",
      "imported":false,
      "kmsKey":"",
      "kubernetesVersion":"1.18",
      "loggingTypes":[
         
      ],
      "nodeGroups":[
         {
            "desiredSize":2,
            "diskSize":20,
            "ec2SshKey":"instance-key",
            "gpu":false,
            "imageId":"",
            "instanceType":"t3.medium",
            "labels":{
               
            },
            "launchTemplate":null,
            "maxSize":2,
            "minSize":2,
            "nodegroupName":"test-cluster-worker",
            "requestSpotInstances":false,
            "resourceTags":{
               
            },
            "spotInstanceTypes":[
               
            ],
            "subnets":[
               
            ],
            "tags":{
               
            },
            "type":"/v3/schemas/nodeGroup",
            "userData":"",
            "version":"1.18"
         }
      ],
      "privateAccess":false,
      "publicAccess":true,
      "publicAccessSources":[
         
      ],
      "region":"us-east-1",
      "secretsEncryption":false,
      "securityGroups":[
         
      ],
      "serviceRole":"",
      "subnets":[
         3 items"subnet-0cf935de179546f9d",
         "subnet-0baeb2a53b354c305",
         "subnet-015a86fe36eb5a2be"
      ],
      "tags":{
         
      },
      "type":"/v3/schemas/eksClusterConfigSpec"
   },
   "eksStatus":{
      "managedLaunchTemplateID":"",
      "managedLaunchTemplateVersions":null,
      "privateRequiresTunnel":null,
      "securityGroups":null,
      "subnets":null,
      "type":"/v3/schemas/eksStatus",
      "upstreamSpec":{
         "amazonCredentialSecret":"cattle-global-data:cc-twprg",
         "displayName":"test-cluster",
         "imported":false,
         "kmsKey":"",
         "kubernetesVersion":"1.18",
         "loggingTypes":[
            
         ],
         "nodeGroups":[
            
         ],
         "privateAccess":false,
         "publicAccess":true,
         "publicAccessSources":[
            "0.0.0.0/0"
         ],
         "region":"us-east-1",
         "secretsEncryption":false,
         "securityGroups":[
            
         ],
         "serviceRole":"",
         "subnets":[
            3 items"subnet-0cf935de179546f9d",
            "subnet-0baeb2a53b354c305",
            "subnet-015a86fe36eb5a2be"
         ],
         "tags":{
            
         },
         "type":"/v3/schemas/eksClusterConfigSpec"
      },
      "virtualNetwork":""
   },
   "enableClusterAlerting":false,
   "enableClusterMonitoring":false,
   "enableNetworkPolicy":false,
   "fleetWorkspaceName":"fleet-default",
   "id":"c-8bpkc",
   "internal":false,
   "istioEnabled":false,
   "labels":{
      "cattle.io/creator":"norman",
      "fleet.cattle.io/cluster-name":"c-8bpkc"
   },
   "limits":{
      "cpu":"0",
      "memory":"0",
      "pods":"0"
   },
   "links":{
      "apiServices":"…/v3/clusters/c-8bpkc/apiservices",
      "clusterAlertGroups":"…/v3/clusters/c-8bpkc/clusteralertgroups",
      "clusterAlertRules":"…/v3/clusters/c-8bpkc/clusteralertrules",
      "clusterAlerts":"…/v3/clusters/c-8bpkc/clusteralerts",
      "clusterCatalogs":"…/v3/clusters/c-8bpkc/clustercatalogs",
      "clusterLoggings":"…/v3/clusters/c-8bpkc/clusterloggings",
      "clusterMonitorGraphs":"…/v3/clusters/c-8bpkc/clustermonitorgraphs",
      "clusterRegistrationTokens":"…/v3/clusters/c-8bpkc/clusterregistrationtokens",
      "clusterRoleTemplateBindings":"…/v3/clusters/c-8bpkc/clusterroletemplatebindings",
      "clusterScans":"…/v3/clusters/c-8bpkc/clusterscans",
      "etcdBackups":"…/v3/clusters/c-8bpkc/etcdbackups",
      "namespaces":"…/v3/clusters/c-8bpkc/namespaces",
      "nodePools":"…/v3/clusters/c-8bpkc/nodepools",
      "nodes":"…/v3/clusters/c-8bpkc/nodes",
      "notifiers":"…/v3/clusters/c-8bpkc/notifiers",
      "persistentVolumes":"…/v3/clusters/c-8bpkc/persistentvolumes",
      "projects":"…/v3/clusters/c-8bpkc/projects",
      "remove":"…/v3/clusters/c-8bpkc",
      "self":"…/v3/clusters/c-8bpkc",
      "shell":"wss://rancher.haya-ventures.com/v3/clusters/c-8bpkc?shell=true",
      "storageClasses":"…/v3/clusters/c-8bpkc/storageclasses",
      "subscribe":"…/v3/clusters/c-8bpkc/subscribe",
      "templates":"…/v3/clusters/c-8bpkc/templates",
      "tokens":"…/v3/clusters/c-8bpkc/tokens",
      "update":"…/v3/clusters/c-8bpkc"
   },
   "localClusterAuthEndpoint":{
      "enabled":false,
      "type":"/v3/schemas/localClusterAuthEndpoint"
   },
   "name":"test-cluster",
   "nodeCount":0,
   "nodeVersion":0,
   "provider":"",
   "requested":{
      "cpu":"0",
      "memory":"0",
      "pods":"0"
   },
   "state":"updating",
   "transitioning":"error",
   "transitioningMessage":"Waiting for API to be available; InvalidUserData.Malformed: Invalid BASE64 encoding of user data.\n\tstatus code: 400, request id: 7c8c9d45-cd42-4264-b8b6-acee9669557f",
   "type":"cluster",
   "uuid":"c60a9041-f799-4f73-8d0c-97a47fcd6c35",
   "windowsPreferedCluster":false
}
```

Below is the error info from AWS Cloudtrail:

```
{
    "eventVersion": "1.08",
    "userIdentity": {
        "type": "IAMUser",
        "principalId": "A**K",
        "arn": "arn:aws:iam::4**6:user/s**n",
        "accountId": "4**6",
        "accessKeyId": "A**5",
        "userName": "s**n"
    },
    "eventTime": "2021-03-06T16:33:35Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateLaunchTemplate",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "34.204.151.222",
    "userAgent": "aws-sdk-go/1.36.7 (go1.13.8b4; linux; amd64)",
    "errorCode": "Client.InvalidUserData.Malformed",
    "errorMessage": "Invalid BASE64 encoding of user data.",
    "requestParameters": {
        "CreateLaunchTemplateRequest": {
            "LaunchTemplateName": "rancher-managed-lt-test-cluster",
            "LaunchTemplateData": {
                "UserData": "<sensitiveDataRemoved>"
            },
            "TagSpecification": {
                "ResourceType": "launch-template",
                "tag": 1,
                "Tag": {
                    "Value": "do-not-modify-or-delete",
                    "tag": 1,
                    "Key": "rancher-managed-template"
                }
            }
        }
    },
    "responseElements": null,
    "requestID": "29673d65-7336-4326-acb9-0d56705e30c5",
    "eventID": "81a6eb0e-89cd-47da-bef3-625f71473e47",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "managementEvent": true,
    "eventCategory": "Management",
    "recipientAccountId": "4**6"
}
```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):v.2.5.6
- Installation option (single install/HA): HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): EKS
- Machine type (cloud/VM/metal) and specifications (CPU/memory): t3.medium
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T21:51:49Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:15:20Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
Version:          19.03.13
```

**Comment by aaronRancher** [comment]
Tested on Rancher `v2.5.6`. Was unable to reproduce this behavior in the following scenarios:

Scenario 1:
Launch an EKS cluster with default values.

Scenario 2:
Navigate to EKS cluster create.
Keep default values.
Remove all user data from launch template.
Launch cluster.

EKS config of Rancher request:
```
eksConfig: {amazonCredentialSecret: "cattle-global-data:cc-48nrw", displayName: "aaron-test2", imported: false,…}
amazonCredentialSecret: "cattle-global-data:cc-48nrw"
displayName: "aaron-test2"
imported: false
kmsKey: ""
kubernetesVersion: "1.18"
loggingTypes: []
nodeGroups: [{desiredSize: 2, diskSize: 20, ec2SshKey: "", gpu: false, imageId: "", instanceType: "t3.medium",…}]
0: {desiredSize: 2, diskSize: 20, ec2SshKey: "", gpu: false, imageId: "", instanceType: "t3.medium",…}
desiredSize: 2
diskSize: 20
ec2SshKey: ""
gpu: false
imageId: ""
instanceType: "t3.medium"
labels: {}
launchTemplate: null
maxSize: 2
minSize: 2
nodegroupName: "aaron-ng"
requestSpotInstances: false
resourceTags: {}
spotInstanceTypes: []
subnets: []
tags: {}
type: "/v3/schemas/nodeGroup"
userData: ""
version: "1.18"
privateAccess: false
publicAccess: true
publicAccessSources: []
region: "us-west-2"
secretsEncryption: false
securityGroups: []
serviceRole: ""
subnets: []
tags: {}
type: "/v3/schemas/eksClusterConfigSpec"
```

When cluster becomes available, user data is set to `null` after EKS sync:
```
"eksConfig": {
"amazonCredentialSecret": "cattle-global-data:cc-48nrw",
"displayName": "aaron-test2",
"imported": false,
"kmsKey": "",
"kubernetesVersion": "1.18",
"loggingTypes": [ ],
"nodeGroups": [
{
"desiredSize": 2,
"diskSize": 20,
"ec2SshKey": null,
"gpu": false,
"imageId": null,
"instanceType": "t3.medium",
"labels": { },
"launchTemplate": null,
"maxSize": 2,
"minSize": 2,
"nodegroupName": "aaron-ng",
"requestSpotInstances": false,
"resourceTags": { },
"spotInstanceTypes": null,
"subnets": [ 3 items
...
],
"tags": { },
"type": "/v3/schemas/nodeGroup",
"userData": null,
"version": "1.18"
}
],
"privateAccess": false,
"publicAccess": true,
"publicAccessSources": [
"0.0.0.0/0"
],
"region": "us-west-2",
"secretsEncryption": false,
"securityGroups": [ ],
"serviceRole": "",
"subnets": [ ],
"tags": { },
"type": "/v3/schemas/eksClusterConfigSpec"
},
```

**Comment by tovern** [comment]
Having the same trouble here trying to provision an EKS 1.18 Cluster on Rancher 2.5.6 UI. 
`Failed to communicate with cluster: Post "https://xxx.eks.amazonaws.com/api/v1/namespaces": dial tcp xxx:443: i/o timeout; InvalidUserData.Malformed: Invalid BASE64 encoding of user data. status code: 400, request id: xxx`
This was previously working on Rancher 2.5.5, I would suspect its related to the new launch template functionality. I have tried with pretty much every combination of settings but no such luck yet. The Cloudformation stack for the service group gets created successfully but the nodegroup stack doesn't even attempt to create.

**Comment by sowmyav27** [comment]
On 2.5.6

- userdata is empty, deploy an EKS cluster in `us-east-1` region
- Error seen on the cluster `Failed to communicate with cluster: Post "https://.....us-east-1.eks.amazonaws.com/api/v1/namespaces": dial tcp 34.194.125.239:443: i/o timeout; InvalidUserData.Malformed: Invalid BASE64 encoding of user data. status code: 400, request id:....`

**Comment by thedadams** [comment]
For those that are experiencing this problem: it seems that this is a region specific problem on AWS. It should work on `us-east-2` region.

**Comment by aaronRancher** [comment]
Reproduced this issue on Rancher `v2.5.6`
Deploy an EKS cluster with empty user data on `us-east-1` region
![Screen Shot 2021-03-12 at 12 38 09 PM](https://user-images.githubusercontent.com/57409052/111000538-7429b200-833f-11eb-8de7-395d8b6504c2.png)

The bug fix is verified on Rancher `master-head` `version d8bfdc9`
Cluster is provisioned successfully.
Node group config in API.
```
nodeGroups": [
{
"desiredSize": 2,
"diskSize": 20,
"ec2SshKey": null,
"gpu": false,
"imageId": null,
"instanceType": "t3.medium",
"labels": { },
"launchTemplate": null,
"maxSize": 2,
"minSize": 2,
"nodegroupName": "aaron-ng2",
"requestSpotInstances": false,
"resourceTags": { },
"spotInstanceTypes": null,
"subnets": [ 3 items
...
],
"tags": { },
"type": "/v3/schemas/nodeGroup",
"userData": null,
"version": "1.18"
}
],
```

EKS Operator version: `rancher/eks-operator:v1.0.7-rc1`

**Comment by cbron** [comment]
This will officially be released in 2.5.8, but it has also been released out-of-band, meaning any clusters that can access ranchers public system-charts repo (system-library) should receive the fix now.
